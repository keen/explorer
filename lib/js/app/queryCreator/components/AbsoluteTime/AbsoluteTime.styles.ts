import styled from 'styled-components';
import { transparentize } from 'polished';
import { colors } from '@keen.io/colors';

import { inputMixin } from '../Input';

export const TimeRow = styled.div`
  display: flex;
  align-items: center;
`;

export const Container = styled.div`
  display: flex;
  flex-direction: column;

  ${TimeRow} + ${TimeRow} {
    margin-top: 10px;
  }
`;

export const TimeLabel = styled.div`
  font-size: 14px;
  line-height: 17px;
  font-family: Lato Regular, sans-serif;
  color: ${colors.blue[500]};
  display: flex;
  flex-shrink: 0;
  align-items: center;
  margin-right: 10px;
  width: 35px;
`;

export const TimePickerContainer = styled.div`
  font-family: 'Lato Regular', sans-serif;
  flex-basis: 50%;

  .rc-time-picker-clear {
    display: none;
  }

  .rc-time-picker .rc-time-picker-input {
    ${inputMixin()}
  }
`;

export const DatePicker = styled.div`
  font-family: 'Lato Regular', sans-serif;
  flex-basis: 50%;
  margin-right: 5px;

  .SingleDatePickerInput__withBorder {
    border: none;
    border-radius: 0;
  }

  .SingleDatePicker_picker {
    top: 39px!important;
  }

  .CalendarDay__default:hover {
    background: ${transparentize(0.8, colors.green[100])};
    border: 1px solid ${transparentize(0.8, colors.green[100])};
  }

  .DayPicker__withBorder {
    border: solid 1px ${colors.gray[200]};
    box-shadow: 0 10px 24px 0 rgba(29, 39, 41, 0.15);
  }

  .DateInput_input {
    ${inputMixin()}
  }

  .DateInput_fang {
    display: none;
  }

  .CalendarDay__selected, .CalendarDay__selected:active, .CalendarDay__selected:hover {
    background: ${colors.blue[500]};
    border: 1px double ${colors.blue[500]};
`;
