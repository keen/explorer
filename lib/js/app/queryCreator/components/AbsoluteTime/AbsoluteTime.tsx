import React, { FC, useState, useMemo, memo } from 'react';
import moment from 'moment-timezone';
import 'react-dates/initialize';
import TimePicker from 'rc-time-picker';
import { SingleDatePicker } from 'react-dates';

import {
  Container,
  DatePicker,
  TimePickerContainer,
  TimeLabel,
  TimeRow,
} from './AbsoluteTime.styles';

import { use12HoursDateFormat } from './utils/hoursFormat';
import { DATE_FORMAT, TIME_PICKER_CLASS } from './constants';

import text from './text.json';

import { Timeframe, Timezones } from '../../types';

type Props = {
  /** Unique identifer */
  id: string;
  /** Time start */
  start: string;
  /** Time end */
  end: string;
  /** Absolute time change event handler */
  onChange: (time: Timeframe) => void;
  /** Timezone value */
  timezone: Timezones;
};

const AbsoluteTime: FC<Props> = memo(
  ({ id, start, end, timezone, onChange }) => {
    const [startFocus, setStartFocus] = useState(false);
    const [endFocus, setEndFocus] = useState(false);

    const startDate = moment(start).tz(timezone);
    const endDate = moment(end).tz(timezone);

    const use12Hours = useMemo(() => use12HoursDateFormat(), []);

    return (
      <Container data-testid="absolute-time">
        <TimeRow>
          <TimeLabel>{text.startDate}</TimeLabel>
          <DatePicker data-testid="date-start">
            <SingleDatePicker
              date={startDate}
              onDateChange={(valueSelected) => {
                onChange({
                  start: valueSelected.format(),
                  end,
                });
              }}
              focused={startFocus}
              onFocusChange={({ focused }) => setStartFocus(focused)}
              isOutsideRange={() => false}
              id={`${id}_start`}
              numberOfMonths={1}
              displayFormat={DATE_FORMAT}
              hideKeyboardShortcutsPanel
            />
          </DatePicker>
          <TimePickerContainer>
            <TimePicker
              popupClassName={TIME_PICKER_CLASS}
              use12Hours={use12Hours}
              showSecond={false}
              value={startDate}
              placement="bottomLeft"
              onChange={(valueSelected) => {
                onChange({
                  start: valueSelected.format(),
                  end,
                });
              }}
            />
          </TimePickerContainer>
        </TimeRow>
        <TimeRow>
          <TimeLabel>{text.endDate}</TimeLabel>
          <DatePicker data-testid="date-end">
            <SingleDatePicker
              date={endDate}
              onDateChange={(valueSelected) => {
                onChange({
                  start,
                  end: valueSelected.format(),
                });
              }}
              focused={endFocus}
              onFocusChange={({ focused }) => setEndFocus(focused)}
              isOutsideRange={() => false}
              id={`${id}_end`}
              numberOfMonths={1}
              displayFormat={DATE_FORMAT}
              hideKeyboardShortcutsPanel
            />
          </DatePicker>
          <TimePickerContainer>
            <TimePicker
              popupClassName={TIME_PICKER_CLASS}
              use12Hours={use12Hours}
              showSecond={false}
              value={endDate}
              onChange={(valueSelected) =>
                onChange({
                  start,
                  end: valueSelected.format(),
                })
              }
            />
          </TimePickerContainer>
        </TimeRow>
      </Container>
    );
  }
);

AbsoluteTime.displayName = 'AbsoluteTime';

export default AbsoluteTime;
