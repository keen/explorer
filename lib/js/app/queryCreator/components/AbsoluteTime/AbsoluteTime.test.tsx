import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import 'jest-styled-components';

import AbsoluteTime from './AbsoluteTime';

const render = (overProps: any = {}) => {
  const props = {
    id: 'date',
    start: '2020-07-29T12:00:00Z',
    end: '2020-07-30T00:00:00Z',
    timezone: 'UTC',
    onChange: jest.fn(),
    ...overProps,
  };

  const wrapper = rtlRender(<AbsoluteTime {...props} />);

  return {
    props,
    wrapper,
  };
};

test('renders date picker for timeframe start', () => {
  const {
    wrapper: { getByTestId },
  } = render();
  const container = getByTestId('date-start');

  expect(container).toMatchSnapshot();
});

test('renders date picker for timeframe end', () => {
  const {
    wrapper: { getByTestId },
  } = render();
  const container = getByTestId('date-end');

  expect(container).toMatchSnapshot();
});
