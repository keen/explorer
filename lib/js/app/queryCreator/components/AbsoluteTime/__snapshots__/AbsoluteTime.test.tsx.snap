// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders date picker for timeframe end 1`] = `
.c0 {
  font-family: 'Lato Regular',sans-serif;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  margin-right: 5px;
}

.c0 .SingleDatePickerInput__withBorder {
  border: none;
  border-radius: 0;
}

.c0 .SingleDatePicker_picker {
  top: 39px!important;
}

.c0 .CalendarDay__default:hover {
  background: rgba(156,192,162,0.2);
  border: 1px solid rgba(156,192,162,0.2);
}

.c0 .DayPicker__withBorder {
  border: solid 1px #F2F2F2;
  box-shadow: 0 10px 24px 0 rgba(29,39,41,0.15);
}

.c0 .DateInput_input {
  padding: 10px 14px;
  height: 37px;
  width: 100%;
  box-sizing: border-box;
  border: 1px solid rgba(29,39,41,0.5);
  border-radius: 4px;
  outline: none;
  font-size: 14px;
  line-height: 17px;
  font-family: 'Lato Regular',sans-serif;
  color: #27566D;
}

.c0 .DateInput_input:focus {
  box-shadow: 0 0 3px 1px rgba(119,163,187,0.5);
}

.c0 .DateInput_input:disabled {
  border: 1px solid rgba(39,86,109,0.5);
  background: #FFFFFF;
}

.c0 .DateInput_input:disabled::-webkit-input-placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_input:disabled::-moz-placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_input:disabled:-ms-input-placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_input:disabled::placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_fang {
  display: none;
}

.c0 .CalendarDay__selected,
.c0 .CalendarDay__selected:active,
.c0 .CalendarDay__selected:hover {
  background: #27566D;
  border: 1px double #27566D;
}

<div
  class="c0"
  data-testid="date-end"
>
  <div
    class="SingleDatePicker SingleDatePicker_1"
  >
    <div>
      <div
        class="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
      >
        <div
          class="DateInput DateInput_1"
        >
          <input
            aria-describedby="DateInput__screen-reader-message-id_end"
            aria-label="Date"
            autocomplete="off"
            class="DateInput_input DateInput_input_1"
            id="id_end"
            name="id_end"
            placeholder="Date"
            type="text"
            value="2020-07-30"
          />
          <p
            class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
            id="DateInput__screen-reader-message-id_end"
          >
            Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders date picker for timeframe start 1`] = `
.c0 {
  font-family: 'Lato Regular',sans-serif;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  margin-right: 5px;
}

.c0 .SingleDatePickerInput__withBorder {
  border: none;
  border-radius: 0;
}

.c0 .SingleDatePicker_picker {
  top: 39px!important;
}

.c0 .CalendarDay__default:hover {
  background: rgba(156,192,162,0.2);
  border: 1px solid rgba(156,192,162,0.2);
}

.c0 .DayPicker__withBorder {
  border: solid 1px #F2F2F2;
  box-shadow: 0 10px 24px 0 rgba(29,39,41,0.15);
}

.c0 .DateInput_input {
  padding: 10px 14px;
  height: 37px;
  width: 100%;
  box-sizing: border-box;
  border: 1px solid rgba(29,39,41,0.5);
  border-radius: 4px;
  outline: none;
  font-size: 14px;
  line-height: 17px;
  font-family: 'Lato Regular',sans-serif;
  color: #27566D;
}

.c0 .DateInput_input:focus {
  box-shadow: 0 0 3px 1px rgba(119,163,187,0.5);
}

.c0 .DateInput_input:disabled {
  border: 1px solid rgba(39,86,109,0.5);
  background: #FFFFFF;
}

.c0 .DateInput_input:disabled::-webkit-input-placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_input:disabled::-moz-placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_input:disabled:-ms-input-placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_input:disabled::placeholder {
  color: rgba(39,52,55,0.4);
}

.c0 .DateInput_fang {
  display: none;
}

.c0 .CalendarDay__selected,
.c0 .CalendarDay__selected:active,
.c0 .CalendarDay__selected:hover {
  background: #27566D;
  border: 1px double #27566D;
}

<div
  class="c0"
  data-testid="date-start"
>
  <div
    class="SingleDatePicker SingleDatePicker_1"
  >
    <div>
      <div
        class="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
      >
        <div
          class="DateInput DateInput_1"
        >
          <input
            aria-describedby="DateInput__screen-reader-message-id_start"
            aria-label="Date"
            autocomplete="off"
            class="DateInput_input DateInput_input_1"
            id="id_start"
            name="id_start"
            placeholder="Date"
            type="text"
            value="2020-07-29"
          />
          <p
            class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
            id="DateInput__screen-reader-message-id_start"
          >
            Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
